### PR TITLE
[1181] no empty copy phases

### PR DESCRIPTION
**Pull request by kalkwarf**
Resolves https://github.com/tuist/tuist/issues/1181

### Short description 📝

Copy phases with no file lists seem inappropriate. Don't generate them.

### Solution 📦

Using the guidance in [1181], I modified `LinkGenerator.generateLinks` to skip steps if the input list is empty.

### Implementation 👩‍💻👨‍💻

- I created a fixture that creates a resource bundle
- I wrote a test that ensures there are no copy files phases when there are no frameworks or libraries
- I fixed the code so the test passed
- I manually ran the code over the fixture, and visually confirmed the result

image

### CODE REVIEW

**Review by pepibumur — COMMENTED** [review]
Thanks for tackling this @kalkwarf. There seem to be legit test errors on CI most likely related to the changes.
One question that I have is how are those fixtures changes used? Since your changes are easily unit-testable, I'd recommend not adding acceptance tests for this. They have a maintenance cost that I'd avoid if we can test the same thing with unit tests.

Great job! 👏

**Comment by pepibumur** [comment]
@all-contributors add @kalkwarf  for bug

**Comment by kalkwarf** [comment]
I was using the fixtures manually, which in hindsight, probably wasn't the smartest of choices. Let me back that out, and then re-read the documentation. I keep projecting my workflow onto this project, and that's just not going to work.

**Comment by kalkwarf** [comment]
I could use some guidance on how to resolve the failed tests. For example
```
Failing Scenarios:
92
cucumber features/generate-4.feature:3 # Scenario: The project is an iOS application with an incompatible Xcode version (ios_app_with_incompatible_xcode)
93
```
looks easy enough. I did `cd fixtures/ios_app_with_incompatible_xcode`, then ran
```
➤ swift run tuist generate
[151/151] Linking tuist
The project, which only supports the versions of Xcode 3.2.1, is not compatible with your selected version of Xcode, 12.0
Fatal linting issues found
```
which looks like the expected result?

**Comment by kwridan** [comment]
Thanks @kalkwarf - looking at the CI logs, it may be the case that the tests were aborted due to other jobs failing (e.g. the unit tests)


```
/Users/runner/work/tuist/tuist/Sources/TuistGenerator/Generator/LinkGenerator.swift:121:13: error: use of unresolved identifier 'generateCopyProductsdBuildPhase'
1161
        try generateCopyProductsdBuildPhase(path: path,
```

**Comment by kalkwarf** [comment]
> Thanks @kalkwarf - looking at the CI logs, it may be the case that the tests were aborted due to other jobs failing (e.g. the unit tests)
> 
> ```
> /Users/runner/work/tuist/tuist/Sources/TuistGenerator/Generator/LinkGenerator.swift:121:13: error: use of unresolved identifier 'generateCopyProductsdBuildPhase'
> 1161
>         try generateCopyProductsdBuildPhase(path: path,
> ```

🤦

File: Sources/TuistGenerator/Generator/LinkGenerator.swift [hunk 1]
```diff
@@ -93,24 +93,37 @@ final class LinkGenerator: LinkGenerating {
...
+                                     pbxproj: pbxproj,
+                                     fileElements: fileElements)
+        }
+
+        try generateCopyProductsBuildPhase(path: path,
+                                           target: target,
+                                           graph: graph,
+                                           pbxTarget: pbxTarget,
+                                           pbxproj: pbxproj,
+                                           fileElements: fileElements)
```
Comment (kwridan) [inline Sources/TuistGenerator/Generator/LinkGenerator.swift:126]: this looks like a duplicate of https://github.com/tuist/tuist/pull/1822/files#diff-bd14a50b734b61af028ddef07944c580R128

File: Sources/TuistGenerator/Generator/LinkGenerator.swift [hunk 2]
```diff
@@ -93,24 +93,37 @@ final class LinkGenerator: LinkGenerating {
...
-                                 pbxproj: pbxproj,
-                                 fileElements: fileElements)
+        if !linkableModules.isEmpty {
+            try setupSearchAndIncludePaths(target: target,
+                                           pbxTarget: pbxTarget,
+                                           sourceRootPath: sourceRootPath,
+                                           path: path,
+                                           graph: graph,
+                                           linkableModules: linkableModules)
+        }
```
Comment (kwridan) [inline Sources/TuistGenerator/Generator/LinkGenerator.swift:103]: looking through, I think it's safe to continue setting up the search and include paths as those are no ops incase there aren't any to setup and they don't add any extra build phase either way.

**Review by kwridan — COMMENTED** [review]
thanks for pushing this @kalkwarf - spotted a few minor bits

**Comment by pepibumur** [comment]
@kalkwarf would it be possible to re-open this PR by pushing the branch directly to the repository (as opposed to using forks). I'd like to investigate the failing tests further and push the fixes directly to the branch.
Thanks a lot! 🙏

**Comment by kalkwarf** [comment]
> @kalkwarf would it be possible to re-open this PR by pushing the branch directly to the repository (as opposed to using forks). I'd like to investigate the failing tests further and push the fixes directly to the branch.
> Thanks a lot! 🙏

Yes, I'll do that momentarily.

**Comment by kalkwarf** [comment]
Wil re-open from a branch.

**Pull request closed without merging**